src/backend/columnar/columnar_reader.c

```diff
@@ -1042,16 +1042,38 @@ GetClauseVars(List *whereClauseList, int natts)
 	 */
 	int flags = 0;
 	List *vars = pull_var_clause((Node *) whereClauseList, flags);
-	Var **deduplicate = palloc0(sizeof(Var *) * natts);
+	return DeduplicateVarList(vars, natts);
+}
+
+
+/*
+ * DeduplicateVarList takes a list of Var objects, and deduplicates and
+ * sorts them in a new list.
+ *
+ * Note that this function doesn't expect any Var's referencing to a system
+ * column or to an invalid attribute. If given varList contains such a Var,
+ * then throws an error.
+ */
+List *
```

Comment:
Can't you use the output of `ColumnarAttrNeeded()`, which is already deduped (because it's a set)?

Reply:
Yeah, but I didn't want to generate Var's from the attribute set again. 

Indeed, this was the first thing I tried, i.e.: re-generating Var's by using `makeVar()` but then I couldn't figure out first and last args that I should pass to this function, then had some issues with the Var's pulled from subqueries.

Reply:
`varno` should be set to `RelOptInfo->relid` (which is not the OID, it's based on position in the query tree).

`varlevelsup` should be 0. If a Var references another table, it's a way to figure out how many levels up it is in the query (if we are in a subquery). In our case, we are always referencing the same level because it's not even a different table.

I agree that it's not necessarily an improvement. Whichever way comes out cleaner in the end is fine with me -- either deduping the Vars or creating new ones from the bitmapset.

Reply:
> varno should be set to RelOptInfo->relid (which is not the OID, it's based on position in the query tree).

Thank you. Looked into code but still couldn't figure out how can I get the `RelOptInfo` for that, any ideas ?

Reply:
You're right, that's not very convenient to access when already at execution time. We could preserve it in the ColumnarScanState, and then use it. It feels like it might be a layering violation, but I don't see an easier way. Maybe it's better to just stick with the dedupe code, and then account for the whole-row vars.